Rancher UI defined roletemplate rules for nonResourceURLs incorrectly have apiGroups definition present, preventing creation of the role in clusters when granted to users

**What kind of request is this (question/bug/enhancement/feature request):**

Bug

**Steps to reproduce (least amount of steps as possible):**

1. Provision a Rancher v2.4.0+ instance with a single all role node custom cluster (I used https://github.com/superseb/tf-do-rancher2/)
2. Under Global -> Security Roles -> Clusters, click `Add Cluster Role`
3. Add a single Resource Grant with `Get` on the Non-Resource URL `/metrics` and remove the `*` in the API Groups field.
4. Grant this new cluster role to a user on the custom cluster.

**Result:**

The roletemplate generated by the UI has a blank, rather than absent `"apiGroups": [""]`, entry, per the following example:

```json
{

    "administrative": false,
    "annotations": {
        "cleanup.cattle.io/rtUpgradeCluster": "true",
        "lifecycle.cattle.io/create.mgmt-auth-roletemplate-lifecycle": "true"
    },
    "baseType": "roleTemplate",
    "builtin": false,
    "clusterCreatorDefault": false,
    "context": "cluster",
    "created": "2020-12-02T16:07:01Z",
    "createdTS": 1606925221000,
    "creatorId": "user-cpgf6",
    "description": "",
    "external": false,
    "hidden": false,
    "id": "rt-h759b",
    "labels": {
        "cattle.io/creator": "norman"
    },
    "links": {
        "clusterRoleTemplateBindings": "…/v3/clusterRoleTemplateBindings?roleTemplateId=rt-h759b",
        "clusters": "…/v3/clusters?defaultClusterRoleForProjectMembers=rt-h759b",
        "projectRoleTemplateBindings": "…/v3/projectRoleTemplateBindings?roleTemplateId=rt-h759b",
        "remove": "…/v3/roleTemplates/rt-h759b",
        "self": "…/v3/roleTemplates/rt-h759b",
        "update": "…/v3/roleTemplates/rt-h759b"
    },
    "locked": false,
    "name": "kube-apiserver metrics",
    "projectCreatorDefault": false,
    "rules": [
        {
            "apiGroups": [
                ""
            ],
            "nonResourceURLs": [
                "/metrics"
            ],
            "type": "/v3/schemas/policyRule",
            "verbs": [
                "get"
            ]
        }
    ],
    "type": "roleTemplate",
    "uuid": "f12c71c5-2c80-44c9-9878-12a2be88e81c"

}
```

As a result of this incorrectly present `apiGroups` definition, for a `nonResourceUrls` rule the clusterole creation will fail when the custom role is granted to a user, with the following error in the rancher server logs:

```
2020/12/02 16:08:07 [INFO] [mgmt-auth-crtb-controller] Updating clusterRoleBinding clusterrolebinding-bnxwk for cluster membership in cluster c-dpvk7 for subject u-fqrkx
2020/12/02 16:08:07 [INFO] Creating clusterRole for roleTemplate kube-apiserver metrics (rt-h759b).
2020/12/02 16:08:07 [ERROR] ClusterRoleTemplateBindingController c-dpvk7/crtb-25kdp [cluster-crtb-sync] failed with : couldn't ensure roles: couldn't create clusterRole rt-h759b: ClusterRole.rbac.authorization.k8s.io "rt-h759b" is invalid: rules[0].nonResourceURLs: Invalid value: []string{"/metrics"}: rules cannot apply to both regular resources and non-resource URLs
```

Same error occurs if you leave the default API Group `*` definition in the form. Removing the `apiGroups` definition from the rule in the role via the API browser, the role and binding is successfully created in the custom cluster.

**Other details that may be helpful:**

When entering a `Non-Resource URL` in the Add Role UI form, the `Resource` field is automatically disabled. The same should occur the API Groups field and no `apiGroups` array should be submitted to the API by the UI for this rule.

The current workaround is to create the role via the API directly.

**Environment information**
- Reproduced with Rancher v2.4.0, v2.4.10 and v2.5.3

--- a/shell/assets/translations/en-us.yaml
+++ b/shell/assets/translations/en-us.yaml
@@ -5712,6 +5712,7 @@ validation:
       missingResource: You must specify a Resource for each resource grant
       missingApiGroup: You must specify an API Group for each resource grant
       missingOneResource: You must specify at least one Resource, Non-Resource URL or API Group for each resource grant
+      noResourceAndNonResource: Each rule may contain Resources or Non-Resource URLs but not both
   service:
     externalName:
       none: External Name is required on an ExternalName Service.

--- a/shell/components/auth/RoleDetailEdit.vue
+++ b/shell/components/auth/RoleDetailEdit.vue
@@ -2,6 +2,8 @@
 import { MANAGEMENT, RBAC } from '@shell/config/types';
 import CruResource from '@shell/components/CruResource';
 import CreateEditView from '@shell/mixins/create-edit-view';
+import FormValidation from '@shell/mixins/form-validation';
+import Error from '@shell/components/form/Error';
 import { RadioGroup } from '@components/Form/Radio';
 import Select from '@shell/components/form/Select';
 import ArrayList from '@shell/components/form/ArrayList';
@@ -58,9 +60,10 @@ export default {
     Tabbed,
     SortableTable,
     Loading,
+    Error
   },
 
-  mixins: [CreateEditView],
+  mixins: [CreateEditView, FormValidation],
 
   async fetch() {
     // We don't want to get all schemas from the cluster because there are
@@ -109,6 +112,9 @@ export default {
       scopedResources:      SCOPED_RESOURCES,
       defaultValue:         false,
       selectFocused:        null,
+      fvFormRuleSets:       [
+        { path: 'displayName', rules: ['required'] }
+      ],
     };
   },
 
@@ -524,7 +530,8 @@ export default {
     :can-yaml="!isCreate"
     :mode="mode"
     :resource="value"
-    :errors="errors"
+    :errors="fvUnreportedValidationErrors"
+    :validation-passed="fvFormIsValid"
     :cancel-event="true"
     @error="e=>errors = e"
     @finish="save"
@@ -568,6 +575,7 @@ export default {
         name-key="displayName"
         description-key="description"
         label="Name"
+        :rules="{ name: fvGetAndReportPathRules('displayName') }"
       />
       <div
         v-if="isRancherType"
@@ -606,6 +614,11 @@ export default {
           :label="t('rbac.roletemplate.tabs.grantResources.label')"
           :weight="1"
         >
+          <Error
+            :value="value.rules"
+            :rules="fvGetAndReportPathRules('rules')"
+            as-banner
+          />
           <ArrayList
             v-model="value.rules"
             label="Resources"

--- a/shell/models/management.cattle.io.globalrole.js
+++ b/shell/models/management.cattle.io.globalrole.js
@@ -4,7 +4,6 @@ import { CATTLE_API_GROUP, SUBTYPE_MAPPING, CREATE_VERBS } from '@shell/models/m
 import { uniq } from '@shell/utils/array';
 import { get } from '@shell/utils/object';
 import SteveDescriptionModel from '@shell/plugins/steve/steve-description-class';
-import Role from './rbac.authorization.k8s.io.role';
 import { AS, MODE, _CLONE, _UNFLAG } from '@shell/config/query-params';
 
 const BASE = 'user-base';
@@ -16,7 +15,14 @@ const GLOBAL = SUBTYPE_MAPPING.GLOBAL.key;
 
 export default class GlobalRole extends SteveDescriptionModel {
   get customValidationRules() {
-    return Role.customValidationRules();
+    return [
+      {
+        path:       'rules',
+        validators: [`roleTemplateRules:${ this.type }`],
+        nullable:   false,
+        type:       'array',
+      },
+    ];
   }
 
   get details() {
@@ -136,6 +142,15 @@ export default class GlobalRole extends SteveDescriptionModel {
 
   async save() {
     const norman = await this.norman;
+
+    for (const rule of norman.rules) {
+      if (rule.nonResourceURLs.length) {
+        delete rule.resources;
+        delete rule.apiGroups;
+      } else {
+        delete rule.nonResourceURLs;
+      }
+    }
 
     return norman.save();
   }

--- a/shell/models/management.cattle.io.roletemplate.js
+++ b/shell/models/management.cattle.io.roletemplate.js
@@ -3,7 +3,6 @@ import { get } from '@shell/utils/object';
 import { DESCRIPTION } from '@shell/config/labels-annotations';
 import { NORMAN } from '@shell/config/types';
 import SteveDescriptionModel from '@shell/plugins/steve/steve-description-class';
-import Role from './rbac.authorization.k8s.io.role';
 import { AS, MODE, _CLONE, _UNFLAG } from '@shell/config/query-params';
 
 export const CATTLE_API_GROUP = '.cattle.io';
@@ -60,7 +59,14 @@ export const CREATE_VERBS = new Set(['PUT', 'blocked-PUT']);
 
 export default class RoleTemplate extends SteveDescriptionModel {
   get customValidationRules() {
-    return Role.customValidationRules();
+    return [
+      {
+        path:       'rules',
+        validators: [`roleTemplateRules:${ this.type }`],
+        nullable:   false,
+        type:       'array',
+      },
+    ];
   }
 
   get details() {
@@ -184,6 +190,15 @@ export default class RoleTemplate extends SteveDescriptionModel {
 
   async save() {
     const norman = await this.norman;
+
+    for (const rule of norman.rules) {
+      if (rule.nonResourceURLs.length) {
+        delete rule.resources;
+        delete rule.apiGroups;
+      } else {
+        delete rule.nonResourceURLs;
+      }
+    }
 
     return norman.save();
   }

--- a/shell/utils/validators/formRules/__tests__/index.test.ts
+++ b/shell/utils/validators/formRules/__tests__/index.test.ts
@@ -393,6 +393,18 @@ describe('formRules', () => {
     expect(formRuleResult).toStrictEqual(expectedResult);
   });
 
+  it('"roleTemplateRules" : returns correct message when type is RBAC role rule defines both resources and nonResourceURLs', () => {
+    const testValue: [{}] = [
+      {
+        verbs: ['verb1'], resources: ['resource1'], apiGroups: ['apiGroup1'], nonResourceURLs: ['nonResourceUrl1']
+      }
+    ];
+    const formRuleResult = formRules.roleTemplateRules('rbac.authorization.k8s.io.role')(testValue);
+    const expectedResult = JSON.stringify({ message: 'validation.roleTemplate.roleTemplateRules.noResourceAndNonResource' });
+
+    expect(formRuleResult).toStrictEqual(expectedResult);
+  });
+
   it('"roleTemplateRules" : returns correct message when type is RBAC role and value is missing apiGroups', () => {
     const testValue: [{}] = [
       {
@@ -408,7 +420,7 @@ describe('formRules', () => {
   it('"roleTemplateRules" : returns undefined when type is not RBAC role and value contains valid rules', () => {
     const testValue: [{}] = [
       {
-        verbs: ['verb1'], nonResourceURLs: ['nonResourceURL1'], resources: ['resource1'], apiGroups: ['apiGroup1']
+        verbs: ['verb1'], resources: ['resource1'], apiGroups: ['apiGroup1']
       }
     ];
     const formRuleResult = formRules.roleTemplateRules('nonrbactype')(testValue);

--- a/shell/utils/validators/formRules/index.ts
+++ b/shell/utils/validators/formRules/index.ts
@@ -370,6 +370,10 @@ export default function(t: Translation, { key = 'Value' }: ValidationOptions): {
       return t('validation.roleTemplate.roleTemplateRules.missingVerb');
     }
 
+    if (val.some((rule: any) => rule.resources?.length && rule.nonResourceURLs?.length)) {
+      return t('validation.roleTemplate.roleTemplateRules.noResourceAndNonResource');
+    }
+
     if (type === RBAC.ROLE) {
       if (val.some((rule: any) => isEmpty(rule.resources))) {
         return t('validation.roleTemplate.roleTemplateRules.missingResource');

--- a/shell/utils/validators/role-template.js
+++ b/shell/utils/validators/role-template.js
@@ -6,6 +6,10 @@ export function roleTemplateRules(rules = [], getters, errors, validatorArgs = [
     errors.push(getters['i18n/t']('validation.roleTemplate.roleTemplateRules.missingVerb'));
   }
 
+  if (rules.some((rule) => rule.resources?.length && rule.nonResourceURLs?.length)) {
+    errors.push(getters['i18n/t']('validation.roleTemplate.roleTemplateRules.noResourceAndNonResource'));
+  }
+
   if (validatorArgs[0] === RBAC.ROLE) {
     if (rules.some((rule) => isEmpty(rule.resources))) {
       errors.push(getters['i18n/t']('validation.roleTemplate.roleTemplateRules.missingResource'));
@@ -13,7 +17,11 @@ export function roleTemplateRules(rules = [], getters, errors, validatorArgs = [
     if (rules.some((rule) => isEmpty(rule.apiGroups))) {
       errors.push(getters['i18n/t']('validation.roleTemplate.roleTemplateRules.missingApiGroup'));
     }
-  } else if (rules.some((rule) => isEmpty(rule.resources) && isEmpty(rule.nonResourceURLs) && isEmpty(rule.apiGroups))) {
+  } else if (rules.some((rule) => rule.resources?.length && rule.nonResourceUrls?.length)) {
+    errors.push(getters['i18n/t']('validation.roleTemplate.roleTemplateRules.noResourceAndNonResource'));
+  }
+
+  if (rules.some((rule) => isEmpty(rule.resources) && isEmpty(rule.nonResourceURLs) && isEmpty(rule.apiGroups))) {
     errors.push(getters['i18n/t']('validation.roleTemplate.roleTemplateRules.missingOneResource'));
   }
 }

